### PR TITLE
Karmada Operator: support disable  karmada cascading deletion

### PR DESCRIPTION
**Pull request by calvin0327**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
In this scenario: user want to retain karmada instance when delete karmada cr.

we support a label `operator.karmada.io/disable-cascading-deletion` in karmada cr.
`operator.karmada.io/disable-cascading-deletion: false` represents delete karmada instance when delete karmada cr.
`operator.karmada.io/disable-cascading-deletion: true` represents  not to delete karmada instance when delete karmada cr.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Karmada Operator: support disable karmada cascading deletion
```

### CODE REVIEW

**Comment by calvin0327** [comment]
@Poor12 @lonelyCZ /cc

File: operator/pkg/controller/karmada/controller.go [hunk 1]
```diff
@@ -52,28 +57,33 @@ func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Req
...
 	}
 
-	// examine DeletionTimestamp to determine if object is under deletion
-	if karmada.DeletionTimestamp.IsZero() {
-		if err := ctrl.ensureKarmada(ctx, karmada); err != nil {
-			return controllerruntime.Result{}, err
+	// The object is being deleted
+	if !karmada.DeletionTimestamp.IsZero() {
+		value, isExist := karmada.Labels[DisableCascadingDeletionLabel]
+		if !isExist || value == strconv.FormatBool(false) {
```
[the remark this answers is not in the document]
Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go:63]: Good a finding. the description  is wrong, it should be:

operator.karmada.io/disable-cascading-deletion: false represents delete karmada instance when delete karmada cr.
operator.karmada.io/disable-cascading-deletion: true represents not to delete karmada instance when delete karmada cr.

[the remark this answers is not in the document]
Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go:63]: > And what do you think is the default value?

thanks, I thinks it's good suggestion. <br>
the other question:  how do you think about the labe `operator.karmada.io/disable-cascading-deletion`.

[the remark this answers is not in the document]
Comment (Poor12) [reply operator/pkg/controller/karmada/controller.go:63]: By default, I think it should be false.

File: operator/pkg/controller/karmada/controller.go [hunk 2]
```diff
@@ -89,6 +99,11 @@ func (ctrl *Controller) ensureKarmada(ctx context.Context, karmada *operatorv1al
 	// then lets add the finalizer and update the object. This is equivalent
 	// registering our finalizer.
 	updated := controllerutil.AddFinalizer(karmada, ControllerFinalizerName)
+	if _, isExist := karmada.Labels[DisableCascadingDeletionLabel]; !isExist {
```
Comment (Poor12) [inline operator/pkg/controller/karmada/controller.go:102]: I think it does not need to fill in the default label if users not set

**Review by Poor12 — COMMENTED** [review]
Please add a release-note about this new feature.

File: operator/pkg/controller/karmada/controller.go (hunk 2 above)
Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go:102]: if there is not a label in the cr,  should we delete karmada instance or not?

Comment (Poor12) [reply operator/pkg/controller/karmada/controller.go:102]: After I think about it, I think it is acceptable to add a default label.

**Comment by Poor12** [comment]
/lgtm
Please update the release note.

**Comment by calvin0327** [comment]
> /lgtm Please update the release note.

Done, thanks.

**Comment by Poor12** [comment]
/approve
We need to update doc later.

**Comment by calvin0327** [comment]
> /approve We need to update doc later.

That's ok.  
@wlp1153468871 are you in interested for this?

**Pull request merged**